portal-impl/src/com/liferay/portlet/PortletContextImpl.java

```diff
@@ -81,17 +86,17 @@ public ClassLoader getClassLoader() {
 
 	@Override
 	public String getContextPath() {
-		throw new UnsupportedOperationException();
+		return _servletContext.getContextPath();
 	}
 
 	@Override
 	public int getEffectiveMajorVersion() {
-		throw new UnsupportedOperationException();
+		return PortletAppUtil.getSpecMajorVersion(_portlet.getPortletApp());
```

Comment:
@shuyangzhou I don't think it can return null, but this was more for the sake of consistency by always calling PortletAppUtil methods. Brian seemed to be happy with PortletAppUtil.

Reply:
@ngriffin7a I don't quite agree with this part. In case PortletApp can not be null, the code should reflect it. Calling PortletAppUtil implicates it could be null. As a matter of fact, if none of the usage of  PortletAppUtil could yield a null PortletApp, we should remove PortletAppUtil. The real argument here is, for now it is good that you can assert none of them can be null. But later over time, if any regression sneaks in causing a null PortletApp, we will not complain anything but quietly fallback to portlet 2.0 behavior, which could lead to more complex issues.

Reply:
@shuyangzhou How about we simply not use PortletAppUtil for now. Once we get rid of all its usages, then we can delete it.

Reply:
@ngriffin7a sure, in case there is valid case that PortletApp can be null, we should explicitly handle it (hope there won't be a case like that.)